src/platform/backends/shared/base_virtual_machine.h

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2021 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef MULTIPASS_BASE_VIRTUAL_MACHINE_H
+#define MULTIPASS_BASE_VIRTUAL_MACHINE_H
+
+#include <multipass/logging/log.h>
+#include <multipass/ssh/ssh_session.h>
+#include <multipass/utils.h>
+#include <multipass/virtual_machine.h>
+
+#include <fmt/format.h>
+
+#include <QRegularExpression>
+#include <QString>
+
+namespace mp = multipass;
+namespace mpl = multipass::logging;
+namespace mpu = multipass::utils;
+
+namespace multipass
+{
+class BaseVirtualMachine : public VirtualMachine
+{
+public:
+    BaseVirtualMachine(VirtualMachine::State state, const std::string& vm_name) : VirtualMachine(state, vm_name){};
+    BaseVirtualMachine(const std::string& vm_name) : VirtualMachine(vm_name){};
+    BaseVirtualMachine(const BaseVirtualMachine&) = delete;
+    BaseVirtualMachine& operator=(const BaseVirtualMachine&) = delete;
+
+    std::vector<std::string> get_all_ipv4(const SSHKeyProvider& key_provider) override
+    {
+        std::vector<std::string> all_ipv4;
+
+        if (state == State::running)
+        {
+            SSHSession session{ssh_hostname(), ssh_port(), ssh_username(), key_provider};
+
+            auto run_in_vm = [&session](const std::string& cmd) -> std::string {
```

Comment:
Great, thanks! A note on points 4 and 5. When using the current daemon implementation of `run_in_vm()`, we use the same SSH session for all the commands we run. If making one call to the VM's function, we'll be creating many SSH sessions. Does this introduce a considerable overhead?

Reply:
Is there any reason why we shouldn't just pass in the `SSHSession` that the daemon creates?

Reply:
Because passing a session does not guarantee that the session is running on that instance. I thought it was cleaner to pass only the data that the VM does not have.

Reply:
Ok, yeah, I see.

Reply:
Ok, moved `run_in_vm()` to utils, taking a `SSHSession` as argument.